### ISSUE TITLE
Parameterize the root directory of the FTP server for non-anonymous logins

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2179]
|<img alt="cwegrzyn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cwegrzyn reported|
|-|-|
|Trac ID|trac#2179|
|Type|enhancement|
|Created|2006-10-19 15:35:11Z|
|Branch|https://github.com/twisted/twisted/tree/ftprealm-custompath-2179-2|

There is no convenient way to have ```FTPRealm()``` use any home other than ```/home/<avatarID>``` for non-anon logins.

Inside ```FTPRealm.requestAvatar()``` why not change the line :

```
avatar = FTPShell(filepath.FilePath("/home/" + avatarId))
```
to

```
avatar = FTPShell(self.anonymousRoot)
```

This would allow the routine invoking ```FTPRealm()``` to set whatever root directory they wish, rather than something fixed?

<details><summary>Searchable metadata</summary>

```
trac-id__2179 2179
type__enhancement enhancement
reporter__cwegrzyn cwegrzyn
priority__normal normal
milestone__ 
branch__branches_ftprealm_custompath_2179_2 branches/ftprealm-custompath-2179-2
branch_author__exarkun__therve exarkun, therve
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1161272111000000 1161272111000000
changetime__1280879882000000 1280879882000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

(please make sure to use meaningful ticket titles)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It would make more sense to add a new parameter to FTPRealm to allow the location of a user directory to be specified.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22144]) Branching to 'ftprealm-custompath-2179'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [22145]) Add a new method getUserHome, test it.

Refs [#2179](https://github.com/twisted/twisted/issues/2179)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

I didn't add a new parameter but a new method, to be customized in subclasses. I thought that it would be more flexible.

Itamar, willing to review it?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I guess this is nice, but I think we should be careful how much code we add to this realm.  Realms are the things _applications_ are supposed to implement.  It's neat to have a very featureful realm in twisted.protocols.ftp, but it's really just a toy.  After all, it can only do IFTPShell.  If you want to handle another interface, you either have to use multiple portals or stop using this realm.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks like ```requestAvatar``` was doing something stupid before: ```filepath.FilePath("/home/" + avatarId)``` is a bit risky, particularly since the requirement for ```avatarId``` to not begin with "../" isn't documented anywhere.  The new ```getUserHome``` method should create a ```FilePath``` for **/home** and then use ```child``` on the ```avatarId```.  This will impose a restriction against nuls and slashes on ```avatarId```, which should be documented.

_Alternatively_, the home directory should be whatever path is in **/etc/passwd** for that username.  This is a bit fishy, since the avatarId may not even correspond to a unix system account - but then, you can say the same thing for using a directory beneath **/home** in the first place.  Using **/etc/passwd** is just the right way to make the wrong assumption. ;)

Having said all that, it now strikes me that splitting this logic out into a unix-specific subclass, mixin, or helper object (consider a callable object passed to ```__init__``` which will be invoked with the ```avatarId``` - ```FilePath('/home/').child``` would work nicely).  Then someone can make an OS X version and a Windows version.  The ftp ```makeService``` could even try to pick the right one.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [22254]) Pass the method in init, rename it.

Refs [#2179](https://github.com/twisted/twisted/issues/2179)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [22270]) Pass a string instead.

Refs [#2179](https://github.com/twisted/twisted/issues/2179)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Anything else I can do to help on this one?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

See also [#4494](https://github.com/twisted/twisted/issues/4494), where someone complained that the current logic doesn't work on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [29620]) Branching to 'ftprealm-custompath-2179-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I implemented something like what I described above, but with subclasses instead of passing callables around.  I also implemented `SystemFTPRealm`, which means this branch also resolved [#4494](https://github.com/twisted/twisted/issues/4494).

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftprealm-custompath-2179-2)

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen<br>TimAllen set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|TimAllen set owner to @exarkun<br>TimAllen set status to `new`|
|-|-|

Looks good, but there's buildbot failures.

On `winxp32-py2.5-select` [as seen here](http://buildbot.twistedmatrix.com/builders/winxp32-py2.5-select/builds/263):
```
[ERROR]: twisted.test.test_ftp.SystemFTPRealmTests.test_getHomeDirectory

Traceback (most recent call last):
  File "C:\twistedbot\winxp32-py2.5-select\Twisted\twisted\test\test_ftp.py", line 2117, in test_getHomeDirectory
    home = realm.getHomeDirectory(user)
  File "C:\twistedbot\winxp32-py2.5-select\Twisted\twisted\protocols\ftp.py", line 1965, in getHomeDirectory
    raise cred_error.UnauthorizedLogin()
twisted.cred.error.UnauthorizedLogin: 
```

On `winxp32-py2.6-select` [as seen here](http://buildbot.twistedmatrix.com/builders/winxp32-py2.6-select/builds/657):
```
[FAIL]: twisted.test.test_ftp.SystemFTPRealmTests.test_noSuchUser

Traceback (most recent call last):
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\test\test_ftp.py", line 2129, in test_noSuchUser
    self.assertRaises(UnauthorizedLogin, realm.getHomeDirectory, user)
twisted.trial.unittest.FailTest: UnauthorizedLogin not raised (FilePath('C:\\Documents and Settings\\bdddf3cf') returned)
```

It's kind of confusing that the two (active) Win32 builders didn't fail the same tests, but suspicious that all the Win32 builders failed in *some* way.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

Okay.  I tried getting this to work on Windows, but it's hard.  So the tests skip on Windows now, and I'm about to file a new ticket specifically for supporting Windows in `SystemFTPRealm`.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftprealm-custompath-2179-2)

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|Screwtape set owner to @exarkun|
|-|-|

Looks good to me. +1 for merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

That ticket I filed is [#4611](https://github.com/twisted/twisted/issues/4611).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [29908]) Merge ftprealm-custompath-2179-2

Author: therve, exarkun
Reviewer: exarkun, screwtape
Fixes: [#2179](https://github.com/twisted/twisted/issues/2179)
Fixes: [#4494](https://github.com/twisted/twisted/issues/4494)
Refs: [#4611](https://github.com/twisted/twisted/issues/4611)

Factor the user home directory selection logic into a helper method of `FTPRealm`
and move the rest of the realm logic into a new `BaseFTPRealm` class.  Add a new
realm, `SystemFTPRealm`, which (on POSIX) can select home directories by looking
at the system user database.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
